posts/2018-05-09-tagless-final-streaming.md

```diff
@@ -0,0 +1,172 @@
+
+### Algebra definition
+
+Consider the following algebra:
+
+```scala
+trait ItemRepository[F[_]] {
+  def findAll: F[List[Item]]
+  def find(name: ItemName): F[Option[Item]]
+  def save(item: Item): F[Unit]
+  def remove(name: ItemName): F[Unit]
+}
+```
+
+Let's go through each method's definition:
+
+- `findAll` needs to return many Items, obtainable inside a context: `F[List[Item]]`.
+- `find` might or might not return an Item inside a context: `F[Option[Item]]`.
+- `save` and `remove` will perform some actions without returning any actual value: `F[Unit]`.
+
+Everything is clear and you might have seen this kind of pattern before, so let's create an interpreter for it:
+
+```scala
+import cats.effect.Sync
+
+// Pretend this is the real implementation... What matters for the example are the types.
+class PostgreSQLItemRepository[F[_]](implicit F: Sync[F]) extends ItemRepository[F] {
+  private val mockItem = Item(ItemName("laptop"), BigDecimal(1500))
+
+  override def findAll: F[List[Item]] = F.pure(List(mockItem))
+  override def find(name: ItemName): F[Option[Item]] = F.pure(Some(mockItem))
+  override def save(item: Item): F[Unit] = F.unit
+  override def remove(name: ItemName): F[Unit] = F.unit
+}
+
+```
+
+Here we have a fake `PostgreSQL` interpreter for the `ItemRepository` algebra, but just pretend that is a real implementation.
+
+One of the most popular DB libraries in the Typelevel ecosystem is [Doobie](http://tpolecat.github.io/doobie/), defined as `A principled JDBC layer for Scala`. And it comes with one super powerful feature: it supports `Streaming` results, since it's built on top of [fs2](https://functional-streams-for-scala.github.io/fs2/).
+
+Now it could be very common to have a huge amount of `Item`s in our DB that a `List` will not fit into memory and / or it will be a very expensive operation. So we might want to stream the results of `findAll` instead of have them all in memory on a `List`, making `Doobie` a great candidate for the job. But wait... We have a problem now. Our `ItemRepository` algebra has fixed the definition of `findAll` as `F[List[Item]]` so we won't be able to create an interpreter that returns a streaming result instead.
+
+### Rethinking our algebra
+
+We should think about abstracting over that `List` and two of the most common abstractions that immediately come to mind are `Foldable` and `Traverse`. But although these typeclasses are very useful, they are not enough to represent a stream of values, so we should come up with a better abstraction.
+
+Well, it seems that our options are either adding another higher-kinded parameter `G[_]` to our algebra or just define an abstract member `G[_]`. So let's go with the first one:
+
+```scala
+trait ItemRepository[F[_], G[_]] {
```

Comment:
Would be interesting to compare this to `trait ItemRepository[F[_], S[_[_], _]] { def findAll: S[F, Item] }`

Reply:
At first sight, the problem I see with this signature is how to create an interpreter where `findAll: List[Item]` or `findAll: F[List[Item]]`? Seems to be more specific to the shape of `fs2.Stream`. By having `F[_], G[_]` I can fix `G` to `Stream[F, ?]` and it also works for types of kind `* -> *` such as `List`, `Vector`, etc.

Reply:
`Stream[Id, ?]`

Reply:
`Stream[Id, ?]`

Reply:
That's an interesting perspective. I'll mention it in this post as another option for the final user, but I'd prefer to avoid diving into more details and leave it for another post :smiley:

Reply:
:) Sounds good to me. I think what bother me about F and G is that their relationship is completely unconstrained. I want to know that F is my effect type and S is my stream type.

Reply:
Sure, in the algebra there's nothing saying that whenever `G[_]` is a `Stream[_[_], _]` the first parameter of that stream should be `F[_]` but I think that could be seen as an advantage as well. The idea of tagless final algebras is to know very little and almost nothing about the final types.

Anyway, I like this design alternative and I'll probably explore it further. Just added a section in this post with it, thanks!